### ISSUE TITLE
Inductor's CppCodecache's initialisation affects the CPU Complex / double's accuracy

### ISSUE THREAD

**Issue by AnthonyBarbier** [pytorch/pytorch#136273]
### 🐛 Describe the bug

For some really weird reason I can't explain importing inductor in an eager mode complex test causes inaccuracies in complex numbers arithmetic, after a lot of time trying to cut down the reproducer I've narrowed it down to the `cdll.LoadLibrary()` call, here is the reproducer inside `CppCodeCache.load("")`:

(this is based on `test_complex_div_underflow_overflow`)

```python
import torch
import os
import pathlib
from ctypes import cdll

if int(os.getenv("NO_LOAD", "0")) == 0:
    lib = pathlib.Path(os.getenv("LIB",""))
    if not lib.is_file():
        from torch._inductor.codecache import CppCodeCache
        path = CppCodeCache.load("")._name
        raise ValueError(f"export LIB=\"{path}\" and re-run.")

    cdll.LoadLibrary(lib)

device="cpu"
dtype=torch.complex128
finfo = torch.finfo(dtype)
nom = torch.tensor([complex(finfo.min / 2, finfo.min / 2)], dtype=dtype, device=device)
denom = torch.tensor([complex(finfo.min / 2, finfo.min / 2)], dtype=dtype, device=device)
res = (nom / denom).item()
# Expected complex(1.0, 0.0)
assert res.real > 0.5, res
print("OK!")
```

Steps:
```
python3 -m venv venv
. venv/bin/activate
pip3 install --pre torch==2.5.0.dev20240823  --index-url https://download.pytorch.org/whl/nightly/cpu
# Make sure inductor is clean
rm -rfv /tmp/torchinductor*
# Run the reproducer once
python3 repro.py
# Set LIB according to the reproducer's output:
export LIB=....
python3 repro.py
```

Will fail with
```
Traceback (most recent call last):
  File "/nethome/anthonyb/tmp/repro.py", line 22, in <module>
    assert res.real > 0.5, res
AssertionError: -0j

```
Re run with `NO_LOAD`:
```sh
NO_LOAD=1 python3 repro.py
```
And it passes.

This occurs on Ubuntu 22.04 with both python 3.10 and 3.12 (Both on x86_64 and aarch64)
I've tried on ArchLinux with python 3.10 and 3.12 and it didn't happen with either.
So maybe it's something to do with GCC? 

I tried to dig into the `Complex`'s logic and added some debug prints in `complex.h`:
```diff
diff --git a/c10/util/complex.h b/c10/util/complex.h
index af810a780d..2c9d639511 100644
--- a/c10/util/complex.h
+++ b/c10/util/complex.h
@@ -1,6 +1,7 @@
 #pragma once
 
 #include <complex>
+#include <iostream>
 
 #include <c10/macros/Macros.h>
 
@@ -236,13 +237,14 @@ struct alignas(sizeof(T) * 2) complex {
     return *this;
   }
 
+#define PP(a) std::cerr<<"["<<__LINE__<<"]"<<#a<<" = "<<a<<std::endl;
 #ifdef __APPLE__
 #define FORCE_INLINE_APPLE __attribute__((always_inline))
 #else
 #define FORCE_INLINE_APPLE
 #endif
   template <typename U>
-  constexpr FORCE_INLINE_APPLE complex<T>& operator/=(const complex<U>& rhs)
+  FORCE_INLINE_APPLE complex<T>& operator/=(const complex<U>& rhs)
       __ubsan_ignore_float_divide_by_zero__ {
     // (a + bi) / (c + di) = (ac + bd)/(c^2 + d^2) + (bc - ad)/(c^2 + d^2) i
     // the calculation below follows numpy's complex division
@@ -250,32 +252,49 @@ struct alignas(sizeof(T) * 2) complex {
     T b = imag_;
     U c = rhs.real();
     U d = rhs.imag();
+    std::cerr<<"a = "<<a<<" b = "<<b<<" c = "<<c<<" d = "<<d<<std::endl;
 
 #if defined(__GNUC__) && !defined(__clang__)
     // std::abs is already constexpr by gcc
     // std::abs is already constexpr by gcc
     auto abs_c = std::abs(c);
     auto abs_d = std::abs(d);
+    PP(abs_c);
+    PP(abs_d);
 #else
+    sdfsdfsd
     auto abs_c = c < 0 ? -c : c;
     auto abs_d = d < 0 ? -d : d;
 #endif
 
     if (abs_c >= abs_d) {
       if (abs_c == 0 && abs_d == 0) {
+        std::cerr<<__FILE__<<":"<<__LINE__<<std::endl;
         /* divide by zeros should yield a complex inf or nan */
         real_ = a / abs_c;
         imag_ = b / abs_d;
+        PP(real_);
+        PP(imag_);
       } else {
+        std::cerr<<__FILE__<<":"<<__LINE__<<std::endl;
         auto rat = d / c;
         auto scl = 1.0 / (c + d * rat);
+        PP(rat);
+        PP(scl);
         real_ = (a + b * rat) * scl;
         imag_ = (b - a * rat) * scl;
+        PP(real_);
+        PP(imag_);
       }
     } else {
+      std::cerr<<__FILE__<<":"<<__LINE__<<std::endl;
       auto rat = c / d;
       auto scl = 1.0 / (d + c * rat);
+      PP(rat);
+      PP(scl);
       real_ = (a * rat + b) * scl;
       imag_ = (b * rat - a) * scl;
+      PP(real_);
+      PP(imag_);
     }
     return *this;
   }
```
Everything matches at the beginning:
```
a = -8.98847e+307 b = -8.98847e+307 c = -8.98847e+307 d = -8.98847e+307
[261]abs_c = 8.98847e+307
[262]abs_d = 8.98847e+307
pytorch/c10/util/complex.h:278
[281]rat = 1
```

The difference occurs in the `scl` calculation: ```[282]scl = -0``` vs ```[282]scl = -5.56268e-309```

How can the codecache loading cause this? No idea :sweat_smile: 

### Versions

/nethome/anthonyb/tmp/venv_dbg/lib/python3.10/site-packages/torch/_subclasses/functional_tensor.py:271: UserWarning: Failed to initialize NumPy: No module named 'numpy' (Triggered internally at ../torch/csrc/utils/tensor_numpy.cpp:84.)
  cpu = _conversion_method_template(device=torch.device("cpu"))
Collecting environment information...
PyTorch version: 2.5.0.dev20240823+cpu
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.4 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: 14.0.0-1ubuntu1.1
CMake version: version 3.22.1
Libc version: glibc-2.35

Python version: 3.10.12 (main, Mar 22 2024, 16:50:05) [GCC 11.4.0] (64-bit runtime)
Python platform: Linux-5.15.0-116-generic-x86_64-with-glibc2.35
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                         x86_64
CPU op-mode(s):                       32-bit, 64-bit
Address sizes:                        48 bits physical, 48 bits virtual
Byte Order:                           Little Endian
CPU(s):                               60
On-line CPU(s) list:                  0-59
Vendor ID:                            AuthenticAMD
Model name:                           AMD EPYC 7742 64-Core Processor
CPU family:                           23
Model:                                49
Thread(s) per core:                   2
Core(s) per socket:                   30
Socket(s):                            1
Stepping:                             0
BogoMIPS:                             4491.56
Flags:                                fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm rep_good nopl cpuid extd_apicid tsc_known_freq pni pclmulqdq ssse3 fma cx16 sse4_1 sse
4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm cmp_legacy svm cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw topoext perfctr_core ssbd ibrs ibpb stibp vmmcall fsgsbase tsc_adjust bmi1 avx2 smep bmi2 rdseed adx smap clflushop
t clwb sha_ni xsaveopt xsavec xgetbv1 clzero xsaveerptr wbnoinvd arat npt lbrv nrip_save tsc_scale vmcb_clean pausefilter pfthreshold v_vmsave_vmload vgif umip rdpid arch_capabilities
Virtualization:                       AMD-V
Hypervisor vendor:                    KVM
Virtualization type:                  full
L1d cache:                            1.9 MiB (30 instances)
L1i cache:                            1.9 MiB (30 instances)
L2 cache:                             15 MiB (30 instances)
L3 cache:                             16 MiB (1 instance)
NUMA node(s):                         2
NUMA node0 CPU(s):                    0-29
NUMA node1 CPU(s):                    30-59
Vulnerability Gather data sampling:   Not affected
Vulnerability Itlb multihit:          Not affected
Vulnerability L1tf:                   Not affected
Vulnerability Mds:                    Not affected
Vulnerability Meltdown:               Not affected
Vulnerability Mmio stale data:        Not affected
Vulnerability Reg file data sampling: Not affected
Vulnerability Retbleed:               Mitigation; untrained return thunk; SMT enabled with STIBP protection
Vulnerability Spec rstack overflow:   Mitigation; safe RET
Vulnerability Spec store bypass:      Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:             Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:             Mitigation; Retpolines; IBPB conditional; STIBP always-on; RSB filling; PBRSB-eIBRS Not affected; BHI Not affected
Vulnerability Srbds:                  Not affected
Vulnerability Tsx async abort:        Not affected

Versions of relevant libraries:
[pip3] torch==2.5.0.dev20240823+cpu
[conda] Could not collect

cc @ezyang @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire

**Comment by ZailiWang** [comment]
Hi, I'm not reproducing the assertion failure with latest nightly. Would you upgrade and verify again?

**Comment by ezyang** [comment]
Someone is probably changing the fp mode...

**Comment by isuruf** [comment]
Fixed in #136466
